status/controller.go

```diff
@@ -65,41 +46,53 @@ func (c *Controller[T]) Register(_ context.Context, m manager.Manager) error {
 
 func (c *Controller[T]) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
 	o := object.New[T]()
-	gvk := object.GVK(o)
 
 	if err := c.kubeClient.Get(ctx, req.NamespacedName, o); err != nil {
 		if errors.IsNotFound(err) {
-			ConditionCount.DeletePartialMatch(map[string]string{
-				MetricLabelGroup:     gvk.Group,
-				MetricLabelKind:      gvk.Kind,
-				MetricLabelNamespace: req.Namespace,
-				MetricLabelName:      req.Name,
-			})
-			ConditionCurrentStatusSeconds.DeletePartialMatch(map[string]string{
-				MetricLabelGroup:     gvk.Group,
-				MetricLabelKind:      gvk.Kind,
-				MetricLabelNamespace: req.Namespace,
-				MetricLabelName:      req.Name,
-			})
-			TerminationCurrentTimeSeconds.DeletePartialMatch(map[string]string{
-				MetricLabelNamespace: req.Namespace,
-				MetricLabelName:      req.Name,
-				MetricLabelGroup:     gvk.Group,
-				MetricLabelKind:      gvk.Kind,
-			})
-			if deletionTS, ok := c.terminatingObjects.Load(req); ok {
-				TerminationDuration.Observe(time.Since(deletionTS.(*metav1.Time).Time).Seconds(), map[string]string{
-					MetricLabelGroup:     gvk.Group,
-					MetricLabelKind:      gvk.Kind,
-					MetricLabelNamespace: req.Namespace,
-					MetricLabelName:      req.Name,
-				})
-			}
+			c.cleanupMetrics(req, o)
 			return reconcile.Result{}, nil
 		}
 		return reconcile.Result{}, fmt.Errorf("getting object, %w", err)
 	}
 
+	return c.reconcile(req, o)
+}
+
+type GenericObjectController[T client.Object] struct {
+	*Controller[unstructuredAdapter]
+}
+
+func NewGenericObjectController[T client.Object](client client.Client, eventRecorder record.EventRecorder) *GenericObjectController[T] {
+	return &GenericObjectController[T]{
+		Controller: NewController[unstructuredAdapter](client, eventRecorder),
+	}
+}
+
+func (c *GenericObjectController[T]) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		For(object.New[T]()).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10}).
+		Named(fmt.Sprintf("operatorpkg.%s.status", strings.ToLower(reflect.TypeOf(object.New[T]()).Elem().Name()))).
+		Complete(c)
+}
+
+func (c *GenericObjectController[T]) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	unstrcut := lo.Must(NewUnstructuredAdapter(object.New[T]()))
```

Comment:
I think this still contains more logic than my PR had -- is it needed? AFAIK, you can combine the Get logic in both of the controllers -- was there a gap in the changes that were proposed in #90?

Reply:
When I building out the testing for this including the Get call resulted in a `nil` pointer deference when trying to get an unstructured type with the adapted pointer. The client was not able to Get adapted unstructured pointer

Reply:
> The client was not able to Get adapted unstructured pointer

Hmmm -- that doesn't make sense -- I was able to validate this by actually running it directly and this worked consistently for me.

Reply:
I fixed this by changing the UnstructuredAdapter to compose an `unstructured.Unstructured` rather than composing an `*unstructured.Unstructured`. This line is the culprit (https://github.com/kubernetes-sigs/controller-runtime/blob/v0.19.1/pkg/client/fake/client.go#L541) -- which `nils` out everything in the object if you have any pointers -- which was causing the panic